deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,683 @@
+		    (int)args->address->ai_addrlen) == 0) {
+		// success, check if we're the winner.
+		pthread_mutex_lock(&context->winner_mutex);
+		os_event_signal(args->candidate->socket_completed_event);
+		if (os_event_try(context->race_completed_event) == EAGAIN) {
+			// We are the winner.
+			context->socket_fd = args->sockfd;
+			memcpy(&context->winner_addr, args->address->ai_addr,
+			       sizeof(struct addrinfo));
+			signal_end(context);
+		}
+		pthread_mutex_unlock(&context->winner_mutex);
+		goto success;
+	}
+
+failure:
+	// failure, note down the error
+	args->candidate->error = GetSockError();
+	pthread_mutex_lock(&context->winner_mutex);
+	os_event_signal(args->candidate->socket_completed_event);
+
+	// connection candidates may still be getting dispatched, treat as if there's
+	// an active candidate
+	bool active = os_atomic_load_bool(&context->is_starting);
+
+	// check if we are the last worker running. If so, we'll set the error status
+	// and signal the completion event.
+	pthread_mutex_lock(&context->candidate_mutex);
+	for (int i = 0; i < context->candidates.num && !active; i++) {
+		active = os_event_try(context->candidates.array[i]
+					      .socket_completed_event) ==
+			 EAGAIN;
+	}
+	pthread_mutex_unlock(&context->candidate_mutex);
+	pthread_mutex_unlock(&context->winner_mutex);
+	if (active || context->error != 0)
+		// we're not last or there is already an error on the context, let's exit.
+		goto success;
+
+	// Ok, we are last. Coalesce errors and signal completion.
+	if (coalesce_errors(context) == STATUS_SUCCESS)
+		signal_end(context);
+
+success:
+	free(args);
+	return NULL;
+}
+
+static int launch_worker(struct happy_eyeballs_ctx *context,
+			 struct addrinfo *addr)
+{
+#ifdef _WIN32
+	SOCKET fd = WSASocket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP, NULL,
+			      0, WSA_FLAG_OVERLAPPED);
+#else
+	SOCKET fd = socket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP);
+#endif
+	if (fd == INVALID_SOCKET) {
+		context->error = GetSockError();
+		return STATUS_FAILURE;
+	}
+	pthread_mutex_lock(&context->candidate_mutex);
+	struct happy_eyeballs_candidate *candidate =
+		da_push_back_new(context->candidates);
+	candidate->sockfd = fd;
+	struct happy_connect_woker_args *args =
+		(struct happy_connect_woker_args *)malloc(
+			sizeof(struct happy_connect_woker_args));
+	if (args == NULL) {
+		context->error = ENOMEM;
+		context->error_message =
+			"happy-eyeballs: Failed to allocate memory for worker context";
+		pthread_mutex_unlock(&context->candidate_mutex);
+		return STATUS_FAILURE;
+	}
+
+	int result = os_event_init(&candidate->socket_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		// failure to create the socket completed event
+		context->error = result;
+		context->error_message =
+			"happy-eyeballs: Failed to initialize socket completed event";
+		free(args);
+		pthread_mutex_unlock(&context->candidate_mutex);
+		return STATUS_FAILURE;
+	}
+	args->sockfd = (int)fd;
```

Comment:
```suggestion
	args->sockfd = fd;
```
Conceptually this cast should be unnecessary? Though I haven't played with the code locally